crypto/fipsmodule/hmac/hmac.c

```diff
@@ -64,34 +64,154 @@
+  DEFINE_IN_PLACE_METHODS(EVP_sha256(), SHA256);
+  DEFINE_IN_PLACE_METHODS(EVP_sha1(), SHA1);
+  DEFINE_IN_PLACE_METHODS(EVP_sha384(), SHA384);
+  DEFINE_IN_PLACE_METHODS(EVP_sha512(), SHA512);
+  DEFINE_IN_PLACE_METHODS(EVP_md5(), MD5);
+  DEFINE_IN_PLACE_METHODS(EVP_sha224(), SHA224);
+  DEFINE_IN_PLACE_METHODS(EVP_sha512_256(), SHA512_256);
+}
+
+static const HmacMethods *GetInPlaceMethods(const EVP_MD *evp_md) {
+  const struct hmac_method_array_st *method_array = AWSLC_hmac_in_place_methods();
+  const HmacMethods *methods = method_array->methods;
+  for (size_t idx = 0; idx < sizeof(method_array->methods) / sizeof(struct hmac_methods_st); idx++) {
+    if (methods[idx].evp_md == evp_md) {
+      return &methods[idx];
+    }
+  }
+  return NULL;
+}
+
+// ctx->state has the following possible states
+// (Pre/Post conditions):
+// HMAC_STATE_UNINITIALIZED: Uninitialized.
+// HMAC_STATE_INIT_NO_DATA: Initialized with an md and key. No data processed.
+//    This means that if init is called but nothing changes, we don't need to reset our state.
+// HMAC_STATE_IN_PROGRESS: Initialized with an md and key. Data processed.
+//    This means that if init is called we do need to reset state.
+// HMAC_STATE_READY_NEEDS_INIT: Identical to state 1 but API contract requires that Init be called prior to use.
+//    This is an optimization because we can leave the context in a state ready for use after completion.
+// other: Invalid state and likely a result of using unitialized memory. Treated the same as 0.
+//
+// While we are within HMAC methods we allow for the state value and actual state of the context to diverge.
+
+// HMAC_STATE_UNINITIALIZED *MUST* remain `0` so that callers can do `HMAC_CTX ctx = {0};` to get a usable context.
+#define HMAC_STATE_UNINITIALIZED 0
+#define HMAC_STATE_INIT_NO_DATA 1
+#define HMAC_STATE_IN_PROGRESS 2
+#define HMAC_STATE_READY_NEEDS_INIT 3
+
+// Indicates that a context has the md and methods configured and is ready to use
+#define hmac_ctx_is_initialized(ctx) ((HMAC_STATE_INIT_NO_DATA == (ctx)->state || HMAC_STATE_IN_PROGRESS == (ctx)->state))
 
 uint8_t *HMAC(const EVP_MD *evp_md, const void *key, size_t key_len,
               const uint8_t *data, size_t data_len, uint8_t *out,
               unsigned int *out_len) {
-  // We have to verify that all the HMAC services actually succeed before
-  // updating the indicator state, so we lock the state here.
+
+  // While returned new contexts are fully zeroed (and that is what we encourage our callers to do),
+  // we know enough about its inner workings to know that setting the state field to HMAC_STATE_UNINITIALIZED
+  // is sufficient for HMAC_Init_ex to get everything into a properly working state.
+  // So this is yet another micro-optimization of only setting a single byte rather than many
+  // (which will just be overwritten in a moment).
+  HMAC_CTX ctx = {.state = HMAC_STATE_UNINITIALIZED};
```

Comment:
What happens to all the fields that are not included in the initializer? Are they initialized to 0?

Reply:
This object has automatic storage duration. So, I think we want to look at C99 standard 6.7.8.19:
>  The initialization shall occur in initializer list order, each initializer provided for a
particular subobject overriding any previously listed initializer for the same subobject; all
subobjects that are not initialized explicitly shall be initialized implicitly the same as
objects that have static storage duration.

Okay, so back to 6.7.8.10:
> If an object that has automatic storage duration is not initialized explicitly, its value is
indeterminate. If an object that has static storage duration is not initialized explicitly,
then:
— if it has pointer type, it is initialized to a null pointer;
— if it has arithmetic type, it is initialized to (positive or unsigned) zero;
— if it is an aggregate, every member is initialized (recursively) according to these rules;
— if it is a union, the first named member is initialized (recursively) according to these
rules.

Hence, you need to recursively work backwards. But if `0` is `NULL`, then all should end up being 0. 

Would it be easier for you to just zeroise everything at initialisation?

Reply:
Thanks. That's what I thought was happening. I just wanted to confirm, and make sure the semantics that is used here matches what Clang and SAW are using. 

The code would probably be more clear if it explicitly initialized this struct to `0`, but the formal verification handles the code that is currently in the PR just fine. Leaving some fields uninitialized would work, too, but it looks like the `md` field must be initialized to `0` in order for `HMAC_INIT_ex` to work correctly. Otherwise `if (md && ctx->md != md)`  on line 262 may fail because `ctx->md`  happens to equal the supplied pointer.

Reply:
You are correct that if `ctx->md` equals the supplied pointer that `HMAC_INIT_ex` will fail (by returning `0` indicating failure). We should simplify this by just replacing it with `HMAC_CTX ctx = {0};`

Reply:
We already require this for `HMAC_STATE_UNINITIALIZED`:

> // HMAC_STATE_UNINITIALIZED *MUST* remain `0` so that callers can do `HMAC_CTX ctx = {0};` to get a usable context.

You could also statically assert it - if you are paranoid (: